packages/flutter/lib/src/widgets/text_selection.dart

```diff
@@ -800,6 +800,169 @@ class _TextSelectionHandleOverlayState
   }
 }
 
+/// An interface to receive gesture handler from [EditableTextSelectionHandlerProvider].
+abstract class EditableTextGestureDetectorClient {
+  /// Key to underlying [RenderEditable].
+  GlobalKey get editableKey;
+
+  /// Whether force press is enabled.
+  bool get forcePressEnabled;
+
+  /// Whether selection is enabled.
+  bool get selectionEnabled;
+}
+
+/// A provider that provides a basic selection handler for [TextSelectionGestureDetector].
+class EditableTextSelectionHandlerProvider {
+  /// Create a [EditableTextSelectionHandlerProvider].
+  EditableTextSelectionHandlerProvider({
+    @required this.client,
+  }) : assert(client != null);
+
+  /// The client to be provided a handler for.
+  final EditableTextGestureDetectorClient client;
+
+  /// Whether to show the selection tool bar.
+  bool shouldShowSelectionToolbar = true;
+
+  /// The [State] of underlying [EditableText].
+  EditableTextState get editableText => client.editableKey.currentState;
+
+  /// The underlying [RenderEditable].
+  RenderEditable get renderEditable => editableText.renderEditable;
+
+  /// Handler callback that will be passed into [TextSelectionGestureDetector.onTapDown].
+  void onTapDown(TapDownDetails details) {
+    renderEditable.handleTapDown(details);
+    // The selection overlay should only be shown when the user is interacting
+    // through a touch screen (via either a finger or a stylus). A mouse shouldn't
+    // trigger the selection overlay.
+    // For backwards-compatibility, we treat a null kind the same as touch.
+    final PointerDeviceKind kind = details.kind;
+    shouldShowSelectionToolbar =
+        kind == null ||
+            kind == PointerDeviceKind.touch ||
+            kind == PointerDeviceKind.stylus;
+  }
+
+  /// Handler callback that will be passed into [TextSelectionGestureDetector.onForcePressStart].
+  void onForcePressStart(ForcePressDetails details) {
+    if (client.selectionEnabled) {
```

Comment:
Could assert client.forcePressEnabled here, to remind the reader, that it's guaranteed.